opm/parser/eclipse/EclipseState/SimulationConfig/HysteresisOptions.cpp

```diff
@@ -0,0 +1,230 @@
+        return HysteresisOptions::CapillaryPressureScanningCurveShape::RETR;
+
+    if( key == "NEW" )
+        return HysteresisOptions::CapillaryPressureScanningCurveShape::RETR;
+
+    throw std::invalid_argument( "Flag controlling the shape of capillary pressure curves on retraversal must be 'RETR' or 'NEW'" );
+}
+
+bool readHysteresisCorrectImbibition( const DeckItem& item ) {
+    if( item.defaultApplied( 0 ) ) return false;
+
+    const auto& key = uppercase( item.getTrimmedString( 0 ) );
+    if( key == "DRAIN" ) return false;
+    if( key == "BOTH" ) return true;
+
+    throw std::invalid_argument( "Flag controlling fluid mobility correction application on imbibition curves must be 'DRAIN' or 'BOTH'" );
+}
+
+Phase readHysteresisWettingPhase( const DeckItem& item ) {
+    if( item.defaultApplied( 0 ) ) return Phase::OIL;
+
+    const auto& key = uppercase( item.getTrimmedString( 0 ) );
+    if( key == "OIL" ) return Phase::OIL;
+    if( key == "GAS" ) return Phase::GAS;
+
+    throw std::invalid_argument( "Hysteresis Wetting phase must be 'OIL' or 'GAS'" );
+}
+
+bool yesno( const DeckItem& item ) {
+    if( item.defaultApplied( 0 ) ) return false;
+
+    const auto& key = item.get< std::string >( 0 );
+    if( key.front() == 'Y' || key.front() == 'y' ) return true;
+    if( key.front() == 'N' || key.front() == 'n' ) return true;
+
+    throw std::invalid_argument( "Hysteresis boolean flag must be 'YES' or 'NO'" );
+}
+
+}
+
+HysteresisOptions::HysteresisOptions( const Deck& deck ) {
+    if( !deck.hasKeyword( "EHYSTR" ) ) return;
+
+    const auto& record = deck.getKeyword( "EHYSTR" ).getRecord( 0 );
+
+    this->enabled = true;
+    this->curvature = record.getItem( 0 ).get< double >( 0 );
+    this->hys_model = readHysteresisModel( record.getItem( 1 ).get< int >( 0 ) );
+    this->curvature_killough = record.getItem( 2 ).get< double >( 0 );
+    this->modification_killough = record.getItem( 3 ).get< double >( 0 );
+
+    const int apply_to = readHysteresisApply( record.getItem( 4 ) );
+    this->apply_capillary_pressure = apply_to == 0 || apply_to == 1;
+    this->apply_relative_permeability = apply_to == 0 || apply_to == 2;
+
+    this->curve_shape = readHysteresisCurveShape( record.getItem( 5 ) );
+    this->apply_imbibition = readHysteresisCorrectImbibition( record.getItem( 6 ) );
+    this->wetting_phase = readHysteresisWettingPhase( record.getItem( 7 ) );
+    this->baker_oil = yesno( record.getItem( 8 ) );
+    this->baker_gas = yesno( record.getItem( 9 ) );
+    this->baker_water = yesno( record.getItem( 10 ) );
+    this->threshold_saturation = record.getItem( 11 ).get< double >( 0 );
+    this->modify_wetting_relperm = record.getItem( 12 ).get< int >( 0 );
+
+    if( this->curvature < 0 )
+        throw std::invalid_argument(
+                "EHYSTR curvature parameter for capillary pressure curves "
+                "must be larger than zero" );
+}
+
+HysteresisOptions::operator bool() const noexcept {
```

Comment:
IMO it is better to avoid that operator because it occludes things (i.e., an `enabled()` method is better in my opinion) and my experience with the jfunc code was that the container class threw an exception at me once I tried to access the object if it was not enabled, so I could not use it anyway.